components/style/context.rs

```diff
@@ -141,6 +143,423 @@ impl<'a> SharedStyleContext<'a> {
     }
 }
 
+/// The structure holds various intermediate inputs that are eventually used by
+/// by the cascade.
+///
+/// The matching and cascading process stores them in this format temporarily
+/// within the `CurrentElementInfo`. At the end of the cascade, they are folded
+/// down into the main `ComputedValues` to reduce memory usage per element while
+/// still remaining accessible.
+#[derive(Clone)]
+pub struct CascadeInputs {
+    /// The rule node representing the ordered list of rules matched for this
+    /// node.
+    rules: Option<StrongRuleNode>,
```

Comment:
super-nit before landing: Could you add a note to this (and/or to `ComputedValues`) saying that we use an `Option` as an optimization to avoid having extra references to the root rule node, but that we really wouldn't need to?